vmap: Fix bug with x * 0.1

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#43218 vmap: Fix bug with x * 0.1**

Previously, `vmap(lambda x: x * 0.1)(torch.ones(3))` would return a
float64 tensor(!!). This is because there is a subtle bug in the
batching rule: the batching rule receives:
- A batched tensor for x
- a scalar tensor: tensor(0.1, dtype=torch.float64).
The batching rule decides to expand the scalar tensor to be the same
size as x and then multiplies the two tensors, promoting the output to
be a float64 tensor. However, this isn't correct: we should treat the
scalar tensor like a scalar tensor. When adding a FloatTensor to a
Double scalar tensor, we don't promote the type usually.

Another example of a bug this PR fixes is the following:
`vmap(torch.mul)(torch.ones(3), torch.ones(3, dtype=torch.float64))`
Multiplying a scalar float tensor with a scalar double tensor produces a
float tensor, but the above produced a float64 before this PR due to
mistakingly type-promoting the tensors.

Test Plan:
- new test: `pytest test/test_vmap.py -v`
- I refactored some tests a bit.

Differential Revision: [D23195418](https://our.internmc.facebook.com/intern/diff/D23195418)

zou3519 committed:
vmap: Fix bug with x * 0.1

Previously, `vmap(lambda x: x * 0.1)(torch.ones(3))` would return a
float64 tensor(!!). This is because there is a subtle bug in the
batching rule: the batching rule receives:
- A batched tensor for x
- a scalar tensor: tensor(0.1, dtype=torch.float64).
The batching rule decides to expand the scalar tensor to be the same
size as x and then multiplies the two tensors, promoting the output to
be a float64 tensor. However, this isn't correct: we should treat the
scalar tensor like a scalar tensor. When adding a FloatTensor to a
Double scalar tensor, we don't promote the type usually.

Another example of a bug this PR fixes is the following:
`vmap(torch.mul)(torch.ones(3), torch.ones(3, dtype=torch.float64))`
Multiplying a scalar float tensor with a scalar double tensor produces a
float tensor, but the above produced a float64 before this PR due to
mistakingly type-promoting the tensors.

Test Plan:
- new test: `pytest test/test_vmap.py -v`
- I refactored some tests a bit.

[ghstack-poisoned]

cpuhrsch reviewed (dismissed):
Looks good. Nice catch.

ezyang commented on aten/src/ATen/BatchingRegistrations.cpp:
```diff
@@ -56,7 +56,36 @@ Tensor sum_batching_rule(const Tensor& self, IntArrayRef dims, bool keepdim, opt
 }
 
 Tensor mul_batching_rule(const Tensor& self, const Tensor& other) {
-  auto physical_args = BroadcastingVmapTransform::logicalToPhysical({self, other});
```
Don't you have to apply this change to everything, not just mul?

zou3519 replied:
mul is actually the only binary broadcasting pointwise operator that has a batching rule right now. I'm going to add the other ones in a later PR (and they'll have this fix).

ezyang commented on aten/src/ATen/BatchingRegistrations.cpp:
```diff
@@ -56,7 +56,36 @@ Tensor sum_batching_rule(const Tensor& self, IntArrayRef dims, bool keepdim, opt
+    return physical_args[0].newLogicalFromPhysical(result);
+  }
+
+  // Handle type promotion of logically scalar tensors.
```
HMMM. At least for these binary operations, can't you just NOT view the scalar, and pass it in directly? At least in those cases, the normal broadcasting behavior of the operation will work properly, AND you'll have cross-device scalars working correctly too (well, correctly as in, however TensorIterator handles them today).

zou3519 replied:
There are two cases here:

Case 1: multiplication between Tensor and Scalar
`vmap(torch.mul, (0, None))(torch.rand(B0), 1)`
In this case, the strategy of not viewing the scalar and passing it in directly works.

Case 2: multiplication between Tensor and a "logically Scalar Tensor"
`vmap(torch.mul)(torch.rand(B0, 3, dtype=torch.float), torch.rand(B0, dtype=torch.double))`
In this case, the strategy of not viewing the scalar and passing it in directly doesn't work. That's because there isn't a "physical scalar" to be viewed here. 

I could split up those two cases so that the cross-device-scalars work for case 1. However, we'll still need some manual type-promotion code to handle case 2 (or we can just ignore case 2 completely for now and mark it as a known issue). What do you think?

ezyang replied:
I haven't worked out all the details, but it seems to me like the distinguishing characteristic between (1) and (2) is whether or not `is_wrapped_number()` is true or not on the Tensor.

zou3519 replied:
I updated the code to distinguish between case (1) (plus another case that I'll call "case 1.5") and case (2).

In case (2) cross-device scalars still don't work, but it should get encountered less often than case (1)

ezyang commented:
Not so sure about the particular implementation strategy here.

zou3519 committed:
Update on "vmap: Fix bug with x * 0.1"


Previously, `vmap(lambda x: x * 0.1)(torch.ones(3))` would return a
float64 tensor(!!). This is because there is a subtle bug in the
batching rule: the batching rule receives:
- A batched tensor for x
- a scalar tensor: tensor(0.1, dtype=torch.float64).
The batching rule decides to expand the scalar tensor to be the same
size as x and then multiplies the two tensors, promoting the output to
be a float64 tensor. However, this isn't correct: we should treat the
scalar tensor like a scalar tensor. When adding a FloatTensor to a
Double scalar tensor, we don't promote the type usually.

Another example of a bug this PR fixes is the following:
`vmap(torch.mul)(torch.ones(3), torch.ones(3, dtype=torch.float64))`
Multiplying a scalar float tensor with a scalar double tensor produces a
float tensor, but the above produced a float64 before this PR due to
mistakingly type-promoting the tensors.

Test Plan:
- new test: `pytest test/test_vmap.py -v`
- I refactored some tests a bit.

Differential Revision: [D23195418](https://our.internmc.facebook.com/intern/diff/D23195418)

[ghstack-poisoned]

ezyang reviewed (approved):
This is pretty subtle. I think I do agree with you that the right long term fix is factoring TensorIterator some more (or maybe giving you a way to say, "Oy, use scalar style promotion here".)

One thing that is kind of weird is vmapping over a 1D tensor, giving you a bunch of scalars. We're willing to promote scalars because we know it's cheap, but when you're a 1D tensor, it's not that cheap anymore... cc @nairbv

zou3519 commented:
> One thing that is kind of weird is vmapping over a 1D tensor, giving you a bunch of scalars. We're willing to promote scalars because we know it's cheap, but when you're a 1D tensor, it's not that cheap anymore... cc @nairbv

One alternative is to just say we don't support type promotion (including the cross-device behavior) when a user vmaps over a 1D tensor and gives a bunch of scalars. It's possible to detect that and raise a nice error message for it inside the Batching Rule. I don't have a strong preference for either, but I haven't seen any potential use cases where a user vmaps over a 1D tensor like that
